modules/unsupported/mbstyle/src/main/java/org/geotools/mbstyle/MBLayer.java

```diff
@@ -17,79 +17,271 @@
  */
 package org.geotools.mbstyle;
 
+import org.geotools.mbstyle.parse.MBObjectParser;
+import org.geotools.styling.StyleFactory2;
+import org.json.simple.JSONArray;
 import org.json.simple.JSONObject;
+import org.opengis.filter.Filter;
+import org.opengis.filter.FilterFactory2;
 
+/**
+ * MBLayer wrapper (around one of the MBStyle layers).
+ * <p>
+ * All methods act as accessors on provided JSON layer, no other state is maintained. This allows
+ * modifications to be made ceanly with out chance of side-effect.
+ * </p>
+ * <p>
+ * In the normal course of events MBLayer is constructed as a flighweight object by MBStyle to
+ * provide easy access to its layers list.
+ * </p>
+ * 
+ * @author Torben Barsballe (Boundless)
+ */
 public abstract class MBLayer {
+    /** Helper class used to provide json access and expression / filter conversions */
+    final protected MBObjectParser parse;
 
-    protected JSONObject json;
+    /** Shared filter factory */
+    final protected FilterFactory2 ff;
 
-    protected String id;
+    /** Shared style factory */
+    final protected StyleFactory2 sf;
 
-    protected String source;
+    /** JSON layer being wrapped. */
+    final protected JSONObject json;
 
-    protected String sourceLayer;
+    /** field access for "id" key (due to its use in error messages) */
+    final protected String id;
 
-    protected Boolean visibility = true;
+    public MBLayer(JSONObject json) {
+        this(json, new MBObjectParser());
+    }
 
-    public MBLayer(JSONObject json) {        
-        super();
+    public MBLayer(JSONObject json, MBObjectParser parse) {
         this.json = json;
-
+        this.parse = parse;
+        this.ff = parse.getFilterFactory();
+        this.sf = parse.getStyleFactory();
+        
         this.id = (String) json.get("id");
+    }
+    //
+    // Factory methods
+    //
+    
+    /**
+     * Factory method creating the appropriate MBStyle based on the "type" indicated in the layer JSON.
+     */
+    public static MBLayer create(JSONObject layer ){
+        if( layer.containsKey("type") && layer.get("type") instanceof String){
+            String type = (String) layer.get("type");
+            switch( type ){
+            case "line": return new LineMBLayer(layer);
+            case "fill": return new FillMBLayer(layer);
+            case "raster": return new RasterMBLayer(layer);
+            default:
+                throw new MBFormatException(("\"type\" " + type
+                        + " is not a valid layer type. Must be one of: "
+                        + "background, fill, line, symbol, raster, circle, fill-extrusion"));
```

Comment:
This error message might be confusing for now (a "background" type would trigger it) -- but probably not a problem